validate filter length doesn't exceed 256 bytes in subscription

omrikiei opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
This PR validates that the filter length doesn't exceed 256 bytes in a PubSub subscription, on TF it should fail when planning
<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:bug

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
pusbub: added validation to `filter` field for `google_pubsub_subscription`
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

hao-nan-li reviewed (commented):
Looks like this change didn't create properly create the validation function. Could you refer to https://github.com/GoogleCloudPlatform/magic-modules/blob/8bed26c9a024ada0fcdb1ddea3bb01df32028c1c/mmv1/products/securitycenter/FolderCustomModule.yaml#L82 and see if that works?

omrikiei commented:
> Looks like this change didn't create properly create the validation function. Could you refer to
> 
> https://github.com/GoogleCloudPlatform/magic-modules/blob/8bed26c9a024ada0fcdb1ddea3bb01df32028c1c/mmv1/products/securitycenter/FolderCustomModule.yaml#L82
> 
> and see if that works?

I've changed it to what I think would instrument a change, not really sure how to test this locally..

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

hao-nan-li commented:
> > Looks like this change didn't create properly create the validation function. Could you refer to
> > https://github.com/GoogleCloudPlatform/magic-modules/blob/8bed26c9a024ada0fcdb1ddea3bb01df32028c1c/mmv1/products/securitycenter/FolderCustomModule.yaml#L82
> > 
> > and see if that works?
> 
> I've changed it to what I think would instrument a change, not really sure how to test this locally..

This change still did not cause any diff. I think you should make the same change in subscription.yaml as well.

omrikiei commented:
> > > Looks like this change didn't create properly create the validation function. Could you refer to
> > > https://github.com/GoogleCloudPlatform/magic-modules/blob/8bed26c9a024ada0fcdb1ddea3bb01df32028c1c/mmv1/products/securitycenter/FolderCustomModule.yaml#L82
> > > 
> > > and see if that works?
> > 
> > 
> > I've changed it to what I think would instrument a change, not really sure how to test this locally..
> 
> This change still did not cause any diff. I think you should make the same change in subscription.yaml as well.

 I see, I changed the file, would it be generated now?

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-11421-old..auto-pr-11421) ( 1 file changed, 5 insertions(+), 3 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-11421-old..auto-pr-11421) ( 1 file changed, 5 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 49
Passed tests: 48
Skipped tests: 1
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>pubsub</li>

</ul>

</blockquote>
</details>

$\textcolor{green}{\textsf{All tests passed!}}$

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11421/artifacts/d636233a-ae50-45ec-a39d-3b814d47f107/build-log/replaying_test.log)

martinOrigin8Cares commented:
Why using a regex to validate the content, If the check is only length?

ndopj commented:
Hello, this change introduced a breaking behaviour which was not released with new major version.
https://github.com/hashicorp/terraform-provider-google/pull/19131#issuecomment-2324053553
Bellow is copy paste with additional info from link above:  

This is breaking change which was not released under new Major version. The API allows filters to contain new line characters. This is useful when Terraform users use [Heredoc strings](https://developer.hashicorp.com/terraform/language/expressions/strings#heredoc-strings) for attributes expecting string type.

Example for PubSub filter:
```hcl
resource "google_pubsub_subscription" "events" {
  name   = "events"
  topic  = google_pubsub_topic.events.name
  filter = <<-EOT
  (attributes.platform = "web" AND attributes.category = "shopping_cart")
  OR (attributes.platform = "app")
  EOT
}
```
The above filter used to work before the change introduced in this MR. Introduced regex, ^.{1,256}$, used for validation of the filter attribute does not allow new lines to be present in the provided string value.
